Fix pending breakdown interfering with mechanic heading to inspection

zaxcav opened the pull request:
If, when a mechanic is heading to inspect a ride, a ride breakdown is prepared, the ride lifecycle_flags for the inspection are removed and the ride mechanic_status is reset, resulting in the mechanic being "reset" when he arrives at the exit station.  The mechanic is reset by setting the peep state to falling. When the exit station is over water, the mechanic will drown. This appears to be be main cause of falling/drowning mechanics in #7176.

Revise peep behaviour so a ride breakdown does not interfere with mechanics already heading to inspect the ride. A mechanic inspecting a broken down ride will change from inspecting to fixing the ride.

Rename 'loc_992A18' to 'peep_fixing_sub_state_mask' and comment the various fixing sub_states accordingly.

zaxcav commented:
Some additional comments:
I've tested the sample save from #7176 provided by @LordMarcel over a 22 year test period.  With these changes I observed <20 of the initial 200 mechanics drowning over 22 years.
This is a 10% drowning rate compared with 60% drowning rate reported in the issue...
I still need to test the previous behaviour with the same park to get a concrete comparison.

During this 22 year test run the rides in the sample park appeared to be otherwise normally cared for (inspected and repaired as necessary).

So far I've only dug through relevant code in Peep.cpp and Ride.cpp.  I'm aware that some breakdown reasons are handled in Vehicles.cpp - the remaining drowning mechanics may be coming from there.

I also have not yet specifically checked how the change in the mechanic's peep status from inspecting to fixing when inspecting a broken down ride is reflected in the peep viewport/lists/etc.

Gymnasiast commented on src/openrct2/ride/Ride.cpp:
```diff
@@ -2555,8 +2553,14 @@ static void ride_mechanic_status_update(sint32 rideIndex, sint32 mechanicStatus)
     ride = get_ride(rideIndex);
-    switch (mechanicStatus) {
-    case RIDE_MECHANIC_STATUS_UNDEFINED:
+
```
I don't completely understand the changes in this area. Could you explain them a bit?

zaxcav replied:
I can try. All the changes are related through, so I've got to explain the lot.
Previously, the ride_prepare_breakdown() function sets the mechanicStatus (via ride->mechanic_status) to RIDE_MECHANIC_STATUS_UNDEFINED and updates the ride-> lifecycle_flags: the RIDE_LIFECYCLE_DUE_INSPECTION flag is cleared and the RIDE_LIFECYCLE_BREAKDOWN_PENDING is set.
Here in ride_mechanic_status_update(), when the mechanicStatus is RIDE_MECHANIC_STATUS_UNDEFINED the pending breakdown is elevated to a proper breakdown (RIDE_LIFECYCLE_BROKEN_DOWN is set).
The flag changes made in ride_prepare_breakdown are checked in peep_update_heading_to_inspect().  If this happens when the mechanic is in the ride exit, he falls through the floor.

Logically I see no need for ride_prepare_breakdown() to change the RIDE_LIFECYCLE_DUE_INSPECTION flag or to change ride->mechanic_status. A breakdown being prepared is separate from these things.
The code in question for escalating the pending breakdown to a proper breakdown uses this change to ride->mechanic_status as the trigger however. So leaving the ride->mechanic_status as is requires modifying this trigger. The updated conditions for this trigger are that the ride->mechanic_status is one of the three listed: RIDE_MECHANIC_STATUS_UNDEFINED to cover the original case and the other two to cover the modified behaviour (although I'm not certain that RIDE_MECHANIC_STATUS_CALLING is possible ... need to check that) additionally for these two we need to check that there is a breakdown pending but the ride is not yet broken down.

One subtle point here:
I left setting the ride->mechanic_status = RIDE_MECHANIC_STATUS_CALLING only if is currently RIDE_MECHANIC_STATUS_UNDEFINED - this means a mechanic on the way to inspect a ride won't be called when it does breakdown. On the other hand, the ride would then call the nearest mechanic who need not be the same one. Thoughts about this?

Another point that I didn't notice before now:
RIDE_MECHANIC_STATUS_UNDEFINED is not actually set anywhere anymore. I need to add this to peep_update_fixing_sub_state_13().

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -5135,6 +5171,8 @@ static bool peep_update_fixing_sub_state_11(bool firstRun, rct_peep * peep, Ride
 /**
  * rct2: 0x006C1474
+ * sub_state_12: applies to fixing all failures & inspections
+ * - see peep_fixing_sub_state_mask[]
```
Move mechanic to station exit.

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -5185,6 +5223,8 @@ static bool peep_update_fixing_sub_state_12(bool firstRun, rct_peep * peep, Ride
 /**
  * rct2: 0x006C1504
+ * sub_state_13: applies to fixing all failures & inspections
+ * - see peep_fixing_sub_state_mask[]
```
Finish inspection/fixing.

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -5227,6 +5267,8 @@ static bool peep_update_fixing_sub_state_13(bool firstRun, sint32 steps, rct_pee
 /**
  * rct2: 0x006C157E
+ * sub_state_14: applies to fixing all failures & inspections
+ * - see peep_fixing_sub_state_mask[]
```
Leave by entrance/exit

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4738,6 +4754,8 @@ static bool peep_update_fixing_sub_state_0(Ride * ride)
 
 /**
  * rct2: 0x006C0F09
+ * sub_state_1: applies to fixing all vehicle specific breakdown reasons
```
Move to broken down vehicle

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4796,6 +4814,11 @@ static bool peep_update_fixing_sub_state_1(bool firstRun, rct_peep * peep, Ride
  * rct2: 0x006C0FD3
+ * sub_state_2:  applies to fixing vehicle restraints stuck closed.
+ * sub_state_3:  applies to fixing vehicle doors stuck closed.
+ * sub_state_4:  applies to fixing vehicle restrains stuck open.
```
Fix individual vehicle

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4838,6 +4861,8 @@ static bool peep_update_fixing_sub_state_2345(bool firstRun, rct_peep * peep, Ri
 
 /**
  * rct2: 0x006C107B
+ * sub_state_6:  applies fixing to vehicle malfunction.
```
Fix broken vehicle train

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4886,6 +4911,8 @@ static constexpr const CoordsXY _992A3C[] = {
 
 /**
  * rct2: 0x006C1114
+ * sub_state_7: applies to fixing station specific breakdowns: safety cut-out, control failure, inspection.
```
Move to station end

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4949,6 +4976,8 @@ static bool peep_update_fixing_sub_state_7(bool firstRun, rct_peep * peep, Ride
 /**
  * rct2: 0x006C11F5
+ * sub_state_8: applies to fixing station specific breakdowns: safety cut-out, control failure, inspection.
+ * - see peep_fixing_sub_state_mask[]
```
Fix station end

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4977,6 +5006,9 @@ static bool peep_update_fixing_sub_state_8(bool firstRun, rct_peep * peep)
 /**
  * rct2: 0x006C1239
+ * sub_state_9: applies to fixing station specific breakdowns: safety cut-out, control failure, inspection.
+ *              applies to fixing brake failure
```
Move to station start

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -5061,6 +5093,8 @@ static bool peep_update_fixing_sub_state_9(bool firstRun, rct_peep * peep, Ride
 
 /**
  * rct2: 0x006C1368
+ * sub_state_10: applies to fixing station specific breakdowns: safety cut-out, control failure, inspection.
```
Fix station start

zaxcav replied:
@duncanspumpkin Thanks for all these naming suggestions. Very helpful.

duncanspumpkin commented:
@Gymnasiast this isn't pathfinding.

zaxcav commented:
@duncanspumpkin I've renamed the peep_fixing sub-states based on your suggestions. Please check whether you are happy with what I've done.

I'd appreciate it if someone could double check that I haven't made any mistakes with the changes to peep_fixing_sub_state_mask (line 4637 in Peep.cpp).

I've also added the reset of ride->mechanic_status when the peep finishes fixing/inspecting (previously sub_state_13) that I previously noted was missing... this should also be done for all actions that interrupt a mechanic heading to a ride, such as: firing the mechanic, picking up the mechanic, changing the mechanic orders/options (is this applicable?).  If anyone can point me to the code for these things, it will save me time searching for them. Thanks.

Then I need to address the remaining 5 breakdown reasons.

duncanspumpkin commented on src/openrct2/peep/Peep.cpp:
```diff
@@ -4641,59 +4726,67 @@ static void peep_update_fixing(sint32 steps, rct_peep * peep)
+    {
+	// Ride has broken down since Mechanic was called to inspect it.
+	// Mechanic identifies the breakdown and switches to fixing it.
+        peep->state     = PEEP_STATE_FIXING;
```
the ride->mechanic_status needs to be updated no?

Ah wait it doesn't as inspecting and fixing are the same thing in terms of `mechanic_status`

duncanspumpkin commented:
>  firing the mechanic, picking up the mechanic, changing the mechanic orders/options (is this applicable?). If anyone can point me to the code for these things, it will save me time searching for them. Thanks.

That does not need to be done `ride_mechanic_status_update` checks to see if the peep is still valid (Exists, assigned to the ride, status is correct). So we can skip adding any reset code elsewhere.

duncanspumpkin commented:
Argh I've made a mistake you need to set mechanic status to undefined when the if fails.

zaxcav commented:
@duncanspumpkin Your last comment is very unclear to me.

In any case, I've got further fixes related to the other breakdown causes waiting to be commited.
There are still rare cases in which the mechanic has problems but it's been weeks since I've had time to investigate the cause.
I should probably just go ahead and commit what I've got.

duncanspumpkin commented:
Yeah just ignore that. I had another think about it and it wasnt an issue.
